File: AncientMagic/AncientMagic/Samples/MyObject.m
//
//  MyObject.m
//  AncientMagic
//
//  Created by Viktor Radulov on 1/17/21.
//

#import "MyObject.h"

@implementation MyObject

- (NSString *)property1and2 {
    return [self.property1 stringByAppendingString:self.property2];
}

+ (NSSet<NSString *> *)keyPathsForValuesAffectingProperty1and2 {
    return [NSSet setWithObjects:@"property1", @"property2", nil];
}

@end

static const char kAMMyBelowedClassOwnerMyObjectPropertyKVOContext = 0;

@interface MyObjectOwner()

@property MyObject *myObject;

@end

@implementation MyObjectOwner

- (instancetype)init
{
    self = [super init];
    if (self) {
        [_myObject addObserver:self forKeyPath:@"property" options:(NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld) context:(void *)&kAMMyBelowedClassOwnerMyObjectPropertyKVOContext];
    }
    return self;
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if (context == &kAMMyBelowedClassOwnerMyObjectPropertyKVOContext)
    {
        // id newValue = change[NSKeyValueChangeNewKey];
        // id oldValue = change[NSKeyValueChangeOldKey];
        
        // Do observing
    }
    else
    {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
}

- (void)dealloc
{
    [_myObject removeObserver:self forKeyPath:@"property" context:(void *)&kAMMyBelowedClassOwnerMyObjectPropertyKVOContext];
}

@end

File: AncientMagic/AncientMagic/Samples/MyObject.swift
//
//  MyObject.swift
//  AncientMagic
//
//  Created by Viktor Radulov on 1/17/21.
//

import Foundation

class MyObject: NSObject {
    @objc dynamic var property = "test"
    
    @objc dynamic var name = "test"
    @objc dynamic var surname = "test"
    
    @objc dynamic var fullname: String { name + " " + surname }
    
    @objc class func keyPathsForValuesAffectingFullname() -> Set<String> {
        [#keyPath(name), #keyPath(surname)]
    }
}

class MyObjectOwner: NSObject {
    @objc dynamic var myObject = MyObject()
    
    var myObjectObserver: AnyObject?
    override init() {
        myObjectObserver = myObject.observe(\.property, options: [.new, .old]) { myObject, change in
//            let newValue = change.newValue
//            let oldValue = change.oldValue
            
            // Do observing
        }
    }
    deinit {
        myObjectObserver = nil
    }
}
